File: send.py
import sys
sys.path.append('/home/satyajit/home/mavlink_custom_message/generated.py')  # Replace '/path/to/generated' with the actual path to the generated directory

from pymavlink import mavutil
import time
import generated as mavlink  # Import the updated MAVLink module

# Set the ngrok tunnel URL and port
ngrok_url = 'localhost'
ngrok_port = '14550'

# Create a MAVLink connection over UDP via ngrok
master = mavutil.mavlink_connection(f'udpout:{ngrok_url}:{ngrok_port}')

# Function to send a custom MAVLink message with a string parameter
def send_custom_mavlink_message():
    # Define the custom message fields for MY_CUSTOM_COMMAND
    target_system = 0
    target_component = 0
    param1 = "Hello, MAVLink!"  # Example string parameter
    param2 = 1.0
    param3 = 2.0
    param4 = 3.0
    param5 = 4.0
    param6 = 5.0
    param7 = 6.0

    # Encode the string parameter to bytes
    param1_encoded = param1.encode("utf-8")

    # Encode the custom message MY_CUSTOM_COMMAND
    msg1 = mavlink.MAVLink_my_custom_command_message(
        target_system,
        target_component,
        param1_encoded,
        param2,
        param3,
        param4,
        param5,
        param6,
        param7
    )

    # Send the MY_CUSTOM_COMMAND message
    master.mav.send(msg1)
    print(f"MY_CUSTOM_COMMAND sent: {msg1}")

    # Define the custom message fields for ANOTHER_CUSTOM_COMMAND
    target_system = 0
    target_component = 0
    param1 = 10.0
    param2 = 20.0
    param3 = 80.5
    param4 = 78.6

    # Encode the custom message ANOTHER_CUSTOM_COMMAND
    msg2 = mavlink.MAVLink_another_custom_command_message(
        target_system,
        target_component,
        param1,
        param2,
        param3,
        param4
    )

    # Send the ANOTHER_CUSTOM_COMMAND message
    master.mav.send(msg2)
    print(f"ANOTHER_CUSTOM_COMMAND sent: {msg2} \n")

# Main loop to send messages periodically
while True:
    send_custom_mavlink_message()
    time.sleep(2)  # Send every 2 seconds

File: send.c
#include <stdio.h>
#include "generated_dir/custom_message/mavlink.h"

// Declare the send_custom_message function before main
void send_custom_message(mavlink_channel_t chan);

int main() {
    // Initialize MAVLink channel (if needed)
    mavlink_channel_t chan = MAVLINK_COMM_0;

    // Call your send_custom_message function
    send_custom_message(chan);

    return 0;
}

// Define the send_custom_message function after main
void send_custom_message(mavlink_channel_t chan) {
    mavlink_message_t msg;
    uint8_t buf[MAVLINK_MAX_PACKET_LEN];
    mavlink_my_custom_command_t custom_msg;

    custom_msg.target_system = 1;  // Example target system ID
    custom_msg.target_component = 1;  // Example target component ID
    strncpy(custom_msg.param1, "Hello, MAVLink!", sizeof(custom_msg.param1));
    custom_msg.param2 = 3.14;
    custom_msg.param3 = 42.0;
    custom_msg.param4 = 1.618;
    custom_msg.param5 = 2.718;
    custom_msg.param6 = 0.577;
    custom_msg.param7 = 1.414;

    mavlink_msg_my_custom_command_encode(1, 200, &msg, &custom_msg);
    uint16_t len = mavlink_msg_to_send_buffer(buf, &msg);
    // Send the buffer over your communication channel (e.g., serial, UDP, etc.)
    // send_uart_bytes(buf, len);
    printf("Custom message sent\n");
}
